numpy/lib/arraypad.py

```diff
@@ -885,105 +885,74 @@ def _pad_wrap(arr, pad_amt, axis=-1):
-        axis.
-        (constant,) or int is a shortcut for before = after = constant for all
-        axes.
+    x : {None, scalar, array-like}
+        The object to broadcast to the shape (`ndim`, 2). None is only allowed
+        as a special case if `as_index` is True.
+    ndim : int
+        Number of pairs the broadcasted `x` will have.
+    as_index : bool, optional
+        If `x` is not None, try to round each element of `x` to a non-negative
+        integer.
 
     Returns
     -------
-    _validate_lengths : tuple of tuples
-        int                               => ((int, int), (int, int), ...)
-        [[int1, int2], [int3, int4], ...] => ((int1, int2), (int3, int4), ...)
-        ((int1, int2), (int3, int4), ...) => no change
-        [[int1, int2], ]                  => ((int1, int2), (int1, int2), ...)
-        ((int1, int2), )                  => ((int1, int2), (int1, int2), ...)
-        [[int ,     ], ]                  => ((int, int), (int, int), ...)
-        ((int ,     ), )                  => ((int, int), (int, int), ...)
-
+    pairs : nested structure with shape (`ndim`, 2)
+        The broadcasted version of `x`.
+
+    Raises
+    ------
+    TypeError
+        If `as_index` is True and `x` is not None and can't be rounded to an
+        array of integer type.
+        Or if `assert_number` is True and the dtype of `x` is not a subdtype
+        of `np.number`.
+    ValueError
+        If `as_index` is True and `x` contains negative elements.
+        Or if `x` is not broadcastable to the shape (`ndim`, 2).
     """
-    normshp = _normalize_shape(narray, number_elements)
-    for i in normshp:
-        chk = [1 if x is None else x for x in i]
-        chk = [1 if x >= 0 else -1 for x in chk]
-        if (chk[0] < 0) or (chk[1] < 0):
-            fmt = "%s cannot contain negative values."
-            raise ValueError(fmt % (number_elements,))
-    return normshp
+    if as_index and x is None:
+        # Pass through None as a special case for indices
+        return ((None, None),) * ndim
+
+    x = np.asarray(x)
+
+    if x.ndim > 2:
+        # Fail explicitly when `x` already has more dimensions than
+        # desired shape (ndim, 2) accounts for
+        raise ValueError("keyword argument must not have more than 2 "
+                         "dimensions, got {}".format(x.ndim))
+
+    if as_index:
+        x = x.round().astype(np.intp, copy=False)
+
+    if x.size == 1:
+        # Single value case
+        x = x.ravel()  # Reduce superfluous dimensions
+        if as_index and x < 0:
+            raise ValueError("index can't contain negative values")
+        return ((x[0], x[0]),) * ndim
+
+    if x.size == 2 and x.shape != (2, 1):
+        # Pair value case, but except special case when each dimension has a
+        # single value which should be broadcasted to a pair, e.g.
+        # [[1], [2]] -> [[1, 1], [2, 2]]
+        x = x.ravel()
+        if as_index and x[0] < 0 and x[1] < 0:
```

Comment:
Second `and` should be `or` (and add parentheses); note that small mistakes like these are reasons beyond code clarity to use the same code path as much as possible - sometimes a micro-optimization is not really worth it...

Reply:
If `broadcast_to` is really a bottleneck, then it would be better to try and speed it up, rather than generate inconsistency downstream.

Reply:
> note that small mistakes like these are reasons beyond code clarity to use the same code path as much as possible - sometimes a micro-optimization is not really worth it...

Alright, that should be a good lesson for me. :sweat_smile: